cpp/src/groupby/hash/groupby.cu

```diff
@@ -589,13 +556,34 @@ rmm::device_uvector<size_type> extract_populated_keys(Map map,
 std::unique_ptr<table> groupby(table_view const& keys,
                                host_span<aggregation_request const> requests,
                                cudf::detail::result_cache* cache,
-                               bool keys_have_nulls,
-                               null_policy include_null_keys,
+                               bool const keys_have_nulls,
+                               null_policy const include_null_keys,
                                rmm::cuda_stream_view stream,
                                rmm::mr::device_memory_resource* mr)
 {
-  auto d_keys_ptr = table_device_view::create(keys, stream);
-  auto map        = create_hash_map(*d_keys_ptr, keys_have_nulls, include_null_keys, stream);
+  auto const num_keys = keys.num_rows();
+  auto const null_keys_are_equal =
+    include_null_keys == null_policy::INCLUDE ? null_equality::EQUAL : null_equality::UNEQUAL;
+  auto const has_null = nullate::DYNAMIC{keys_have_nulls};
+
+  auto preprocessed_keys = cudf::experimental::row::hash::preprocessed_table::create(keys, stream);
+  cudf::experimental::row::equality::self_comparator comparator(preprocessed_keys);
+  auto row_hash    = cudf::experimental::row::hash::row_hasher(std::move(preprocessed_keys));
```

Comment:
`row_hash` confiscates the content of `preprocessed_keys` before `comparator` is used.

Reply:
> row_hash confiscates the content of preprocessed_keys before comparator is used.

This shouldn't be an issue since the content is shared via a shared pointer

> So don't use std::move here and preprocessed_keys will be shared among them.

Moving the shared pointer (https://github.com/rapidsai/cudf/pull/10770#discussion_r868208284) is a minor improvement so we can avoid expensive reference counting.

Reply:
> row_hash confiscates the content of preprocessed_keys before comparator is used. So don't use std::move here and preprocessed_keys will be shared among them.

This isn't correct. `preprocessed_keys` is a `shared_ptr`, so we aren't confiscating it's managed object, we are just eliminating an extraneous copy of the `shared_ptr` object.